src/Workspaces/CSharp/Portable/VirtualChars/CSharpVirtualCharService.cs

```diff
@@ -0,0 +1,263 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.VirtualChars;
+
+namespace Microsoft.CodeAnalysis.CSharp.VirtualChars
+{
+    [ExportLanguageService(typeof(IVirtualCharService), LanguageNames.CSharp), Shared]
+    internal class CSharpVirtualCharService : AbstractVirtualCharService
+    {
+        public static readonly IVirtualCharService Instance = new CSharpVirtualCharService();
+
+        protected override ImmutableArray<VirtualChar> TryConvertToVirtualCharsWorker(SyntaxToken token)
+        {
+            Debug.Assert(!token.ContainsDiagnostics);
+            if (token.Kind() != SyntaxKind.StringLiteralToken)
+            {
+                return default;
+            }
+
+            return token.IsVerbatimStringLiteral()
+                ? TryConvertVerbatimStringToVirtualChars(token)
+                : TryConvertStringToVirtualChars(token);
+        }
+
+        private ImmutableArray<VirtualChar> TryConvertVerbatimStringToVirtualChars(SyntaxToken token)
+            => TryConvertSimpleDoubleQuoteString(token, "@\"", "\"");
+
+        private ImmutableArray<VirtualChar> TryConvertStringToVirtualChars(SyntaxToken token)
+        {
+            const string StartDelimeter = "\"";
+            const string EndDelimeter = "\"";
+
+            var tokenText = token.Text;
+            if (!tokenText.StartsWith(StartDelimeter) ||
+                !tokenText.EndsWith(EndDelimeter))
+            {
+                Debug.Fail("This should not be reachable as long as the compiler added no diagnostics.");
+                return default;
+            }
+
+            var startIndexInclusive = StartDelimeter.Length;
+            var endIndexExclusive = tokenText.Length - EndDelimeter.Length;
+
+            var result = ArrayBuilder<VirtualChar>.GetInstance();
+            try
+            {
+                var offset = token.SpanStart;
+                for (var index = startIndexInclusive; index < endIndexExclusive;)
+                {
+                    if (tokenText[index] == '\\')
+                    {
+                        if (!TryAddEscape(result, tokenText, offset, index))
+                        {
+                            return default;
```

Comment:
Should we release the `result` array builder here?

Reply:
:)